File: src/simpleEquation.py

from typing import Any, Tuple
import numpy as np

class GA:
    def __init__(self, n_population: int, mu: float = 0.01, seed: Any = 64) -> None:
        self.rng = np.random.default_rng(seed)
        # self.population = self.rng.integers(2, size=(n_population, 5))
        self.population = self.rng.random(size=(n_population, 5)) <= 0.5
        self.mu = mu
    
    def pop2int(self, pop: np.ndarray) -> np.ndarray:
        ''' popを整数値の配列に変換して返す '''
        s = pop.shape
        b2i = 2**np.arange(s[1]-1, -1, -1).reshape((1, -1)).repeat(s[0], axis=0)
        return (pop*b2i).sum(axis=1)

    def fitness(self, pop: np.ndarray) -> np.ndarray:
        #ref: https://living-sun.com/ja/python/718584-binary-numpy-array-to-list-of-integers-python-numpy-binary.html
        int_values = self.pop2int(pop)
        return np.abs(int_values - 5)

    def bestIndividuals(self, pop: np.ndarray, fits: np.ndarray, k: int = 1,) -> Tuple[np.ndarray, np.ndarray]:
        ''' 集団popから上位k個体を抽出して返す '''
        ranks = np.argsort(fits)
        pop_sorted = pop[ranks]
        fits_sorted = fits[ranks]
        return pop_sorted[0:k, :], fits_sorted[0:k]

    def selection(self, pop: np.ndarray, fits: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
        n_population = pop.shape[0]
        next_pop, next_fits = self.bestIndividuals(pop, fits, k=n_population//2)
        return next_pop, next_fits
    
    def crossover(self, pop: np.ndarray) -> np.ndarray:
        n_population, n_gene = pop.shape
        id_all = self.rng.choice(n_population, n_population, replace=False)
        
        n_couple = n_population//2
        id_dad = id_all[:n_couple]
        id_mom = id_all[n_couple:]
        
        offsprings = []
        
        for dad_index, mom_index in zip(id_dad, id_mom):
            dad = pop[dad_index]
            mom = pop[mom_index]
            
            doSwap = self.rng.integers(2, size=n_gene) == 1
            
            boy, girl = dad.copy(), mom.copy()
            boy[doSwap], girl[doSwap] = girl[doSwap], boy[doSwap]
        
            offsprings.append(boy)
            offsprings.append(girl)
        
        return np.vstack((pop, np.array(offsprings)))
    
    def mutation(self, pop: np.ndarray) -> np.ndarray:
        doMutation = self.rng.uniform(0, 1, size=pop.shape) < self.mu
        return np.where(doMutation, 1 - pop, pop)
    
    def solve(self, generation: int) -> int:
        for g in range(generation):
            pop = self.population
            fits = self.fitness(pop)
            
            mean = fits.mean()
            min_ = fits.min()
            print(f"--- Generation: {g+1} ---")
            print(f"\tmean: {mean}")
            print(f"\tmin: {min_}")
            
            pop, fits = self.selection(pop, fits)
            
            pop = self.crossover(pop)
            
            pop = self.mutation(pop)
            
            self.population = pop
        
        best, _ = self.bestIndividuals(pop, fits=self.fitness(pop), k=1)
        return self.pop2int(best)[0]

def main():
    solver = GA(10, 0.01)
    
    ans = solver.solve(100)
    
    print(f"x = {ans}")

if __name__ == '__main__':
    main()

File: src/simpleEquation.jl

# ref: https://qiita.com/AokiMasataka/items/4fe9b2623282238ba6f3

using Random

Random.seed!(0)

# 集団を表す型エイリアス。Int8の二次元配列の意味。各個体の遺伝子は行で表される。
Population = Matrix{Int8}

# 個体を表すエイリアス
Individual = Vector{Int8}

# 個体の評価値をまとめた行列。 n_population × 1 Matrix
EvaluatedIndividuals = Matrix{Int}

"""
    init(genome_length, n_population)

初期集団を生成して返す。
"""
function init(genome_length::Int, n_population::Int):: Population
    population::Population = rand(0:1, n_population, genome_length)
    return population
end

"""
    bin2den(population)

populationを各個体が表す10進数の値の行列にして返す。
"""
function bin2den(population::Population):: EvaluatedIndividuals
    length = size(population)[2]
    # [2^(l-1), ... , 2^1, 2^0]
    binDigit = 2 .^ (length-1:-1:0)'
    return sum(population .* binDigit, dims=2)
end

"""
    evaluate(population)

各個体の評価値を返す。n_population × 1 Matrixを返す。
"""
function evaluate(population::Population):: EvaluatedIndividuals
    return abs.(bin2den(population) .- 5)
end

"""
    select(population, k)

評価の高い個体をk個体選んで返す。
"""
function select(population::Population, k::Int):: Population
    n_population = size(population)[1]
    evaluated:: EvaluatedIndividuals = evaluate(population)
    sorted = population[sortperm(evaluated[:]), :]
    selected = sorted[1:k, :]

    return selected
end

"""
    crossover(x, y)

両親x, yから2個体を新しく作り出して返却する。一様交叉。
"""
function crossover(x::Individual, y::Individual):: Tuple{Individual, Individual}
    boy = similar(x)
    girl = similar(y)

    for i in eachindex(x)
        if rand() >= 0.5
            boy[i] = x[i]
            girl[i] = y[i]
        else
            boy[i] = y[i]
            girl[i] = x[i]
        end
    end

    return boy, girl
end

"""
    mutate!(population, rate)

突然変異を起こす。各遺伝子に確率rateで反転が起きる。
"""
function mutate!(population::Population, μ:: Number) ::Population
    for i in eachindex(population)
        if rand() < μ
            population[i] = 1 - population[i]
        end
    end
    return population
end

"""
    next_generation!(population [, μ])

次世代を生成して返却する。μは突然変異の確率。

# Attention

n_populationは４の倍数でなければならない。
上位N//2が生存し、その集団がN//4カップル作り、各々が2個体を増やすことで集団をN個体に維持する。
"""
function next_generation!(population::Population; μ::Number = 0.01):: Population

    @assert 0 < μ < 1 "rate must be larger than 0 and smaller than 1."

    n_population, l = size(population)
    
    @assert n_population % 4 == 0 "n_population must be shown by 4×N where N is Integer."

    # ÷ は少数切り捨て演算
    k::Int = n_population ÷ 2
    # 自然淘汰
    population = select(population, k)

    # 子孫の集団を宣言
    offsprings = similar(population, k, l)

    # カップルの数
    n_couple = k ÷ 2

    for i in 1:n_couple
        dad = population[i, :]
        mom = population[i+n_couple, :]
        boy, girl = crossover(dad, mom)
        offsprings[i, :] = boy'
        offsprings[i+n_couple, :] = girl'
    end

    # 次世代生成
    population = cat(population, offsprings, dims=1)

    # 突然変異
    population = mutate!(population, μ)

    return population
end

function main()
    genome_length = 5
    n_population = 4 * 3
    mutationRate = 0.01
    generation = 50

    population:: Population = init(genome_length, n_population)

    for i in 1:generation
        population = next_generation!(population; μ=mutationRate)

        evaluated = evaluate(population)

        println("Generation $(i) ------------")
        print("avg: ")
        println(sum(evaluated) / n_population)
        print("min: ")
        println(minimum(evaluated))
        println("")
    end

    best = select(population, 1)
    println("x = ", bin2den(best)[1])

end

main()
